### PR TITLE
Mod/filters layout

### PR DESCRIPTION
**Pull request by pemeraldy**
1. Implemented horizontal display of all services and filters in the tools bar.
2. Handled states of each filter component.

### CODE REVIEW

File: src/views/dashboard/panel/Tool.vue [hunk 1]
```diff
@@ -138,13 +200,19 @@ import { useSelectorStore } from "@/store/modules/selectors";
...
 const { t } = useI18n();
 const dashboardStore = useDashboardStore();
 const selectorStore = useSelectorStore();
 const appStore = useAppStoreWithOut();
+const traceStore = useTraceStore();
 const params = useRoute().params;
-
+const selectedSelector = ref<string>("");
+const showFilter = computed(
+  () => dashboardStore.layout[0]?.activedTabIndex === 2
```
Comment (BFergerson) [inline src/views/dashboard/panel/Tool.vue:213]: You can't hardcode tab indexes like this. People are able to move tabs around and create custom dashboards. What is this trying to achieve?

File: yarn.lock [hunk 2]
```diff
@@ -0,0 +1,13674 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (BFergerson) [inline yarn.lock:1]: Are changes to this file necessary? If not, remove changes

File: src/views/dashboard/related/trace/Detail.vue [hunk 3]
```diff
@@ -155,26 +36,37 @@ import List from "./components/List.vue";
...
-  setup() {
+  setup(props, ctx) {
     const { t } = useI18n();
     const traceStore = useTraceStore();
     const loading = ref<boolean>(false);
     const traceId = ref<string>("");
-    const displayMode = ref<string>("List");
+    const queries = useRoute().query;
+    // console.log("APP::", queries);
+    const isFullVull = computed(() => {
```
Comment (BFergerson) [inline src/views/dashboard/related/trace/Detail.vue:54]: vull is not a real word. Also, remove unnecessary commented code.

File: src/components/Selector.vue [hunk 4]
```diff
@@ -13,7 +13,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. -->
 <template>
-  <el-select
+  <el-select
```
Comment (BFergerson) [inline src/components/Selector.vue:16]:
```suggestion
  <el-select
```

File: src/views/dashboard/controls/Tab.vue [hunk 5]
```diff
@@ -266,7 +266,7 @@ export default defineComponent({
       }
     );
 
-    function clickTabs(e: Event, idx: number) {
+    function clickTabs(e: Event, idx: number) {
```
Comment (BFergerson) [inline src/views/dashboard/controls/Tab.vue:269]:
```suggestion
    function clickTabs(e: Event, idx: number) {
```

File: src/views/dashboard/controls/Trace.vue [hunk 6]
```diff
@@ -56,9 +55,17 @@ const props = defineProps({
 });
 const { t } = useI18n();
 const dashboardStore = useDashboardStore();
+const traceStore = useTraceStore();
+// const traceListActive = ref<boolean>(true);
```
Comment (BFergerson) [inline src/views/dashboard/controls/Trace.vue:59]: Clean up

File: src/views/dashboard/panel/Tool.vue [hunk 7]
```diff
@@ -138,13 +202,22 @@ import { useSelectorStore } from "@/store/modules/selectors";
...
+);
+const showLogHeader = computed(
+  () => dashboardStore.layout[0]?.activedTabIndex === 3
```
Comment (BFergerson) [inline src/views/dashboard/panel/Tool.vue:218]: You can't do this. Users are free to move tabs around. This logic should be based on whether or not the logs component is in use.

File: src/views/dashboard/related/trace/Detail.vue (hunk 3 above)
Comment (pemeraldy) [reply src/views/dashboard/related/trace/Detail.vue:54]: this is a typo, will correct the spelling.

File: src/views/dashboard/panel/Tool.vue (hunk 1 above)
Comment (pemeraldy) [reply src/views/dashboard/panel/Tool.vue:215]: I see what you mean, will use a more dynamic determinant.

File: src/views/dashboard/panel/Tool.vue (hunk 7 above)
Comment (pemeraldy) [reply src/views/dashboard/panel/Tool.vue:218]: Noted

File: src/views/dashboard/controls/Trace.vue (hunk 6 above)
Comment (pemeraldy) [reply src/views/dashboard/controls/Trace.vue]: cleaned up

Comment (pemeraldy) [reply src/views/dashboard/controls/Trace.vue]: cleaned up

File: src/views/dashboard/panel/Tool.vue [hunk 8]
```diff
@@ -138,13 +205,17 @@ import { useSelectorStore } from "@/store/modules/selectors";
...
+import TraceDetailsTools from "./component/TraceDetailsTools";
 const { t } = useI18n();
 const dashboardStore = useDashboardStore();
 const selectorStore = useSelectorStore();
 const appStore = useAppStoreWithOut();
+const traceStore = useTraceStore();
 const params = useRoute().params;
-
+const selectedSelector = ref<string>("");
+const showFilter = computed(() => dashboardStore.currentTabName === "Trace");
```
Comment (BFergerson) [inline src/views/dashboard/panel/Tool.vue:216]: Again, you can't hardcode this logic to tabs. Watch this video: https://youtu.be/Eivyq1rl0Mc

Notice that the traces don't show unless I modify this logic to be true. That is because I'm not using any tabs. This logic must be based on the traces widget being present. Same with the log widget.

Comment (BFergerson) [reply src/views/dashboard/panel/Tool.vue:216]: My guess is that you need to search `DashboardState.layout` for the traces/logs widget. If you find them, then you can add filters. Make sure it still works if you have tabs or if you don't have tabs.

Comment (pemeraldy) [reply src/views/dashboard/panel/Tool.vue:216]: If  I may ask, Is it Ok to base this on the child Items of the current layout if it has the type of trace  as used here
`dashboardStore.setCurrentTabType(extra.children[0].type);` ?

Comment (pemeraldy) [reply src/views/dashboard/panel/Tool.vue:216]: I just watched the video, I think I know how to go about it now, Yeah, the second suggestion you made will be my first approach.

Comment (pemeraldy) [reply src/views/dashboard/panel/Tool.vue:216]: image
I pulled before making a push and got this...

Comment (BFergerson) [reply src/views/dashboard/panel/Tool.vue:216]: I'm not sure why. You can see my commit above doesn't change that.

I would guess you're on the wrong branch or you need to reinstall dependencies.

Comment (pemeraldy) [reply src/views/dashboard/panel/Tool.vue:216]: Ok, Let me try that.

**Comment by BFergerson** [comment]
Two issues with logs:
1. Searching for nothing shouldn't highlight a button. This should be for all buttons. An empty search means that filter is not applied.
2. Include/exclude keywords buttons don't let you input anything

Both of these errors are visible here: https://youtu.be/CdNdPuHytGE

---

One issue with traces:
1. The "Endpoint" dropdown should be a button and placed to the right of the "Services" button. 

This issue is visible here:
Untitled

**Comment by BFergerson** [comment]
Actually, that "Endpoint" dropdown shouldn't be there. If you click the cloud button, that lets you set the endpoint. That's a duplicate endpoint dropdown and it doesn't actually work.

**Comment by pemeraldy** [comment]
Noted

**Comment by pemeraldy** [comment]
> Actually, that "Endpoint" dropdown shouldn't be there. If you click the cloud button, that lets you set the endpoint. That's a duplicate endpoint dropdown and it doesn't actually work.

So do I remove the cloud button then?

**Comment by pemeraldy** [comment]
As for the keyword include/exclude input not showing, that is due to this function in logStore
`async getLogsByKeywords() {
      const res: AxiosResponse = await graphql
        .query("queryLogsByKeywords")
        .params({});

      if (res.data.errors) {
        return res.data;
      }

      this.supportQueryLogsByKeywords = res.data.data.support;
      return res.data;
    },`
    Is it Ok not to show the buttons (include/exclude keywords) if the support value from the endpoint 
 = false ?

**Comment by BFergerson** [comment]
> > Actually, that "Endpoint" dropdown shouldn't be there. If you click the cloud button, that lets you set the endpoint. That's a duplicate endpoint dropdown and it doesn't actually work.
> 
> So do I remove the cloud button then?

No, the cloud button is the correct working one.



> As for the keyword include/exclude input not showing, that is due to this function in logStore
> 
> ```
> async getLogsByKeywords() {
>       const res: AxiosResponse = await graphql
>         .query("queryLogsByKeywords")
>         .params({});
> 
>       if (res.data.errors) {
>         return res.data;
>       }
> 
>       this.supportQueryLogsByKeywords = res.data.data.support;
>       return res.data;
>     }
> ```
> 
> Is it Ok not to show the buttons (include/exclude keywords) if the support value from the endpoint is equal false?

Correct, if it doesn't work don't show the buttons.

**Comment by pemeraldy** [comment]
Ok, I will implement these shortly.

**Comment by pemeraldy** [comment]
I noticed the loader on the Trace component does not show when fetching data... 
Will want to look into it, is that Ok?

**Comment by BFergerson** [comment]
> @BFergerson You can give the current changes a test. I am confident all edge cases have been looked into.

You're very close. Remaining issues:

### The endpoint dropdown is still visible at http://localhost:8080/dashboard/GENERAL/Service/c3Bw.1/General-Service

This is the same issue I said above. I said that the cloud button shows the endpoint dropdown but actually the endpoint dropdown never disappears. Pressing any button shows the endpoint dropdown. Like so:
endpoint-dropdown

Instead, it should be a button, like the rest.

### Log tags are not removed when you press cancel

tags-not-removed

### Instance button's contents not aligned

not-aligned

**Comment by BFergerson** [comment]
> I noticed the loader on the Trace component does not show when fetching data... Will want to look into it, is that Ok?

No, don't worry about that. One of your future tasks will be to remove the loaders.

File: src/views/dashboard/controls/Topology.vue [hunk 9]
```diff
@@ -61,6 +61,7 @@ function editConfig() {
   dashboardStore.setConfigPanel(true);
   dashboardStore.selectWidget(props.data);
 }
+
 </script>
```
Comment (BFergerson) [inline src/views/dashboard/controls/Topology.vue:65]:
```suggestion
</script>
```

**Comment by pemeraldy** [comment]
> > @BFergerson You can give the current changes a test. I am confident all edge cases have been looked into.
> 
> You're very close. Remaining issues:
> 
> ### The endpoint dropdown is still visible at http://localhost:8080/dashboard/GENERAL/Service/c3Bw.1/General-Service
> This is the same issue I said above. I said that the cloud button shows the endpoint dropdown but actually the endpoint dropdown never disappears. Pressing any button shows the endpoint dropdown. Like so: endpoint-dropdown
> 
> Instead, it should be a button, like the rest.
> 
> ### Log tags are not removed when you press cancel
> tags-not-removed
> 
> ### Instance button's contents not aligned
> not-aligned

Wow, looking into it now.

File: src/store/modules/trace.ts [hunk 10]
```diff
@@ -48,6 +51,9 @@ export const traceStore = defineStore({
     services: [{ value: "0", label: "All" }],
     instances: [{ value: "0", label: "All" }],
     endpoints: [{ value: "0", label: "All" }],
+    displayMode: "List",
+    currentView: "traceList",
+    activeFilter: "titi",
```
Comment (BFergerson) [inline src/store/modules/trace.ts:56]: What does `titi` mean?

File: src/store/modules/log.ts [hunk 11]
```diff
@@ -47,7 +47,7 @@ export const logStore = defineStore({
       queryDuration: useAppStoreWithOut().durationTime,
       paging: { pageNum: 1, pageSize: 15, needTotal: true },
     },
-    supportQueryLogsByKeywords: true,
+    supportQueryLogsByKeywords: false,
```
Comment (BFergerson) [inline src/store/modules/log.ts:50]: Why hardcode false?

File: src/views/dashboard/panel/Tool.vue [hunk 12]
```diff
@@ -138,13 +205,17 @@ import { useSelectorStore } from "@/store/modules/selectors";
...
+import TraceDetailsTools from "./component/TraceDetailsTools";
 const { t } = useI18n();
 const dashboardStore = useDashboardStore();
 const selectorStore = useSelectorStore();
 const appStore = useAppStoreWithOut();
+const traceStore = useTraceStore();
 const params = useRoute().params;
-
+const selectedSelector = ref<string>("");
+const showFilter = computed(() => dashboardStore.showTraceTools);
```
Comment (BFergerson) [inline src/views/dashboard/panel/Tool.vue:216]: Should be called `showTraceHeader` to match `showLogHeader`

File: src/views/dashboard/related/trace/Filter.vue [hunk 13]
```diff
@@ -200,6 +337,7 @@ async function queryTraces() {
   }
 }
 function changeField(type: string, opt: any) {
+  activeFilter;
```
Comment (BFergerson) [inline src/views/dashboard/related/trace/Filter.vue:340]: What does this do?

File: src/store/modules/dashboard.ts [hunk 14]
```diff
@@ -37,9 +37,10 @@ interface DashboardState {
   activedGridItem: string;
   durationTime: Duration;
   selectorStore: any;
-  showTopology: boolean;
```
Comment (BFergerson) [inline src/store/modules/dashboard.ts:40]: Why remove this?

File: src/store/modules/trace.ts (hunk 10 above)
Comment (pemeraldy) [reply src/store/modules/trace.ts:56]: I am shocked that is even there, it a typo.. Removing it now.

File: src/views/dashboard/panel/Tool.vue (hunk 12 above)
Comment (pemeraldy) [reply src/views/dashboard/panel/Tool.vue:216]: Noted

File: src/store/modules/log.ts (hunk 11 above)
Comment (pemeraldy) [reply src/store/modules/log.ts:50]: Its dependent on the result from the query handled by this function ` getLogsByKeywords()`

File: src/views/dashboard/related/trace/Filter.vue (hunk 13 above)
Comment (pemeraldy) [reply src/views/dashboard/related/trace/Filter.vue:359]: Nothing actually, I removed that already before pushing don't know it still here. I will cross check again

File: src/store/modules/dashboard.ts (hunk 14 above)
Comment (pemeraldy) [reply src/store/modules/dashboard.ts:40]: Please let me cross check

**Comment by pemeraldy** [comment]
You can run another test now.

File: src/views/dashboard/related/trace/Filter.vue (hunk 13 above)
Comment (BFergerson) [reply src/views/dashboard/related/trace/Filter.vue:359]: This is still present

**Comment by BFergerson** [comment]
Also, it looks like trace tree view no longer works:
notree

**Comment by pemeraldy** [comment]
> @pemeraldy, two remaining issues:
> 
> 1. If you activate multiple filters, you should get multiple buttons highlighted green
> 2. If you cancel one filter, it shouldn't cancel all the other filters.
> 
> Both of these issues are visible here: two-issues

Ok, I will implement it I already did that for logHeader. Though I remember pointing this out on upwork chat earlier and you were like, its not a priority.

**Comment by pemeraldy** [comment]
> Also, it looks like trace tree view no longer works: notree

I think this issue had been right from the creation of the jetbrain-ui branch the  of this issue... Ive always wanted to bring it up but I get busy implementing other features.

**Comment by BFergerson** [comment]
I don

> > @pemeraldy, two remaining issues:
> > 
> > 1. If you activate multiple filters, you should get multiple buttons highlighted green
> > 2. If you cancel one filter, it shouldn't cancel all the other filters.
> > 
> > Both of these issues are visible here: two-issues
> 
> Ok, I will implement it I already did that for logHeader. Though I remember pointing this out on upwork chat earlier and you were like, its not a priority.

I just said not to implement it if it doesn't exist. As in, the goal isn't to add new features. Just match current features with the new UI.



> > Also, it looks like trace tree view no longer works: notree
> 
> I think this issue had been right from the creation of the jetbrain-ui branch... I've always wanted to bring it up but I get busy implementing other features.
> 
> I am trying to cross-check this now though but something seems to be wrong with the server at the moment, I keep getting `500 error status` image image

The demo server appears to be up for me. It's restarted every couple of hours so maybe you just needed to refresh a few times while it was rebooting. Either way, please add a new comment if you want to make sure I get notified about something. I don't get emails when you edit comments and emails are how I keep up-to-date with what you're doing.

**Comment by BFergerson** [comment]
In the future, if you want to use a different demo server, revert this commit: https://github.com/sourceplusplus/skywalking-booster-ui/pull/8/commits/9e1427a9d8329bcdfd0fb71b46485ba63913d80f

You shouldn't need it but it's there if you do.

**Comment by pemeraldy** [comment]
> In the future, if you want to use a different demo server, revert this commit: [9e1427a](https://github.com/sourceplusplus/skywalking-booster-ui/commit/9e1427a9d8329bcdfd0fb71b46485ba63913d80f)
> 
> You shouldn't need it but it's there if you do.

Ok, noted and thanks.

**Comment by pemeraldy** [comment]
@BFergerson Good day,
I have implemented the changes requested.

**Comment by BFergerson** [comment]
> @BFergerson Good day, I have implemented the changes requested.

I'm still seeing the same issues with using multiple filters for traces. Example:
multifilter
> I chose the error status and POST:/users filter, I then removed the error status filter but the POST:/users filter also got reset.


Also, there is a new issue:
autoclose
> The filters sometimes automatically close. Not sure why

**Comment by pemeraldy** [comment]
> > @BFergerson Good day, I have implemented the changes requested.
> 
> I'm still seeing the same issues with using multiple filters for traces. Example: multifilter
> 
> > I chose the error status and POST:/users filter, I then removed the error status filter but the POST:/users filter also got reset.
> 
> Also, there is a new issue: autoclose
> 
> > The filters sometimes automatically close. Not sure why

But I was thinking that when you filter for errors, and later remove errors. This change should reflect in the result. 
Just to understand better, do you mean that when you cancel an active filter, the change should not reflect in the current result of traceList ?

**Comment by pemeraldy** [comment]
Or do I put it this way, the removal of an active filter, should not take effect on the list of traceList initially filtered?

**Comment by BFergerson** [comment]
I'm not too sure what you mean but I'm referring to when I said:
> If you cancel one filter, it shouldn't cancel all the other filters.

In my screencast above, I activated two filters:
- Status: Error
- Endpoint: POST:/users

It works up to that point. I then cancel one of the filters:
- Status: Error

This means I should still be filtering on:
- Endpoint: POST:/users

But that is not the case.

**Comment by pemeraldy** [comment]
Oh, sorry. I think I see what you are saying. Canceling a filter looks like it refreshes the list. I will look into it. Something must have been wrongly implemented.

**Comment by BFergerson** [comment]
Canceling a filter should refresh the list. It shouldn't cancel all the other filters though. Just the one being canceled.

**Comment by pemeraldy** [comment]
Understood

**Comment by pemeraldy** [comment]
Please don't test yet, I am still on it.

**Comment by BFergerson** [comment]
> Please don't test yet, I am still on it.

Thanks for the reminder but I've been waiting till you actually say to test before testing.

**Comment by pemeraldy** [comment]
Oh, ok. cool.

**Comment by pemeraldy** [comment]
@BFergerson You can please test now.

**Comment by BFergerson** [comment]
It looks like canceling a filter doesn't fully work for logs:
instance-reset
> I canceled trace id but instance gets reset too

**Comment by pemeraldy** [comment]
[Will look into it...](https://github.com/sourceplusplus/skywalking-booster-ui/pull/8#issuecomment-1127394745)
Oh, I will look into it soon.

**Comment by pemeraldy** [comment]
Hell @BFergerson, So I did a thorouh check on this one :) . You can test the log section again, I think it works pretty well now.

**Comment by BFergerson** [comment]
@pemeraldy, I'll check it out soon. If you'd like to get started on your next task, check out: https://github.com/sourceplusplus/skywalking-booster-ui/issues/9. Make sure to use a new branch.

**Comment by pemeraldy** [comment]
Ok, I will check it out.

**Comment by pemeraldy** [comment]
Is it ok to checkout from the current branch as It's up to date with the progress so far?

**Comment by BFergerson** [comment]
> Is it ok to checkout from the current branch as It's up to date with the progress so far?

Of course

**Comment by BFergerson** [comment]
The duration filter does not work:
duration

And the trace id copy button does not work:
image
> Also, please remove the blue around it. Make it look more like a tiny button

**Comment by pemeraldy** [comment]
> The duration filter does not work: duration
> 
> And the trace id copy button does not work: image
> 
> > Also, please remove the blue around it. Make it look more like a tiny button

I will look into it.

**Comment by pemeraldy** [comment]
Looking into this duration filter, I think there was really no exact implementation for the minimum and maximum feature( Correct me if I'm wrong though). 
I will be making couple of changes to the `app.ts` file to see that it works as expected.

**Comment by BFergerson** [comment]
@pemeraldy, looks like it might be a bug in the original implementation.

It does appear to work if I modify this:
https://github.com/pemeraldy/skywalking-booster-ui/blob/0c5a108d2030dc92340ddd06d780b0afed8e4ac8/src/views/dashboard/related/trace/Filter.vue#L363-L364


To:
```
    minTraceDuration: 200,
    maxTraceDuration: 300,
```

---

So it appears it will work if it's fed the proper values.

**Comment by pemeraldy** [comment]
@BFergerson You can test now, please.

**Comment by BFergerson** [comment]
@pemeraldy, everything seems to be working. I will do a more thorough check soon. In the meantime, please contribute your trace duration fix to https://github.com/apache/skywalking-booster-ui/. I'm sure they will appreciate your contribution.

**Comment by pemeraldy** [comment]
> @pemeraldy, everything seems to be working. I will do a more thorough check soon. In the meantime, please contribute your trace duration fix to https://github.com/apache/skywalking-booster-ui/. I'm sure they will appreciate your contribution.

Ok, glad to know it works fine. I will be on the lookout for any updates though and yeah, will make the contribution to https://github.com/apache/skywalking-booster-ui/
